File: backend/oracle/oracle.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"

	"github.com/joho/godotenv"
)

const aviationStackAPIURL = "https://api.aviationstack.com/v1/flights"

type FlightData struct {
	FlightStatus   string `json:"flight_status"`
	Departure      string `json:"departure"`
	Arrival        string `json:"arrival"`
	FlightNumber   string `json:"flight_number"`
	AirlineName    string `json:"airline_name"`
	AircraftType   string `json:"aircraft_type"`
	DepartureTime  string `json:"departure_time"`
	ArrivalTime    string `json:"arrival_time"`
	DepartureCity  string `json:"departure_city"`
	ArrivalCity    string `json:"arrival_city"`
}

func FetchFlightData(flightID string) (*FlightData, error) {
	err := godotenv.Load("../../.env")
	if err != nil {
		log.Fatalf("Error loading .env file: %v", err)
	}

	apiKey := os.Getenv("AVIATION_STACK_API_KEY")
	if apiKey == "" {
		return nil, fmt.Errorf("API key is missing")
	}

	url := fmt.Sprintf("%s?access_key=%s&flight_iata=%s", aviationStackAPIURL, apiKey, flightID)

	resp, err := http.Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch flight data: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("API request failed with status code %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %v", err)
	}

	// fmt.Println("Raw API Response: ", string(body))

	var flightDataResponse map[string]interface{}
	if err := json.Unmarshal(body, &flightDataResponse); err != nil {
		return nil, fmt.Errorf("failed to parse response JSON: %v", err)
	}

	flightDetails, ok := flightDataResponse["data"].([]interface{})
	if !ok || len(flightDetails) == 0 {
		return nil, fmt.Errorf("no flight data found for flight ID %s", flightID)
	}

	flight := flightDetails[0].(map[string]interface{})
	flightData := &FlightData{
		FlightStatus:   flight["flight_status"].(string),
		Departure:      flight["departure"].(map[string]interface{})["estimated"].(string),
		Arrival:        flight["arrival"].(map[string]interface{})["estimated"].(string),
		FlightNumber:   flight["flight"].(map[string]interface{})["iata"].(string),
		AirlineName:    flight["airline"].(map[string]interface{})["name"].(string),
		AircraftType:   flight["aircraft"].(map[string]interface{})["iata"].(string),
		DepartureTime:  flight["departure"].(map[string]interface{})["estimated"].(string),
		ArrivalTime:    flight["arrival"].(map[string]interface{})["estimated"].(string),
		DepartureCity:  flight["departure"].(map[string]interface{})["airport"].(string),
		ArrivalCity:    flight["arrival"].(map[string]interface{})["airport"].(string),
	}

	return flightData, nil
}

func main() {
	flightID := "AA100" // Example Flight ID (American Airlines Flight 100)

	flightData, err := FetchFlightData(flightID)
	if err != nil {
		log.Fatalf("Error fetching flight data: %v", err)
	}

	fmt.Printf("Flight Status: %s\n", flightData.FlightStatus)
	fmt.Printf("Departure: %s at %s\n", flightData.DepartureCity, flightData.DepartureTime)
	fmt.Printf("Arrival: %s at %s\n", flightData.ArrivalCity, flightData.ArrivalTime)
	fmt.Printf("Flight Number: %s\n", flightData.FlightNumber)
	fmt.Printf("Airline: %s\n", flightData.AirlineName)
	fmt.Printf("Aircraft: %s\n", flightData.AircraftType)
}

File: backend/oracle/oracle.py
import os
import requests
from dotenv import load_dotenv

load_dotenv()

AVIATION_STACK_API_URL = "https://api.aviationstack.com/v1/flights"

class FlightData:
    def __init__(self, flight_status, departure, arrival, flight_number, airline_name,
                 aircraft_type, departure_time, arrival_time, departure_city, arrival_city):
        self.flight_status = flight_status
        self.departure = departure
        self.arrival = arrival
        self.flight_number = flight_number
        self.airline_name = airline_name
        self.aircraft_type = aircraft_type
        self.departure_time = departure_time
        self.arrival_time = arrival_time
        self.departure_city = departure_city
        self.arrival_city = arrival_city

def fetch_flight_data(flight_id):
    api_key = os.getenv("AVIATION_STACK_API_KEY")
    if not api_key:
        raise ValueError("API key is missing")

    url = f"{AVIATION_STACK_API_URL}?access_key={api_key}&flight_iata={flight_id}"

    response = requests.get(url)
    if response.status_code != 200:
        raise Exception(f"API request failed with status code {response.status_code}")

    flight_data_response = response.json()
    flight_details = flight_data_response.get("data", [])

    if not flight_details:
        raise ValueError(f"No flight data found for flight ID {flight_id}")

    flight = flight_details[0]

    flight_data = FlightData(
        flight_status=flight["flight_status"],
        departure=flight["departure"]["estimated"],
        arrival=flight["arrival"]["estimated"],
        flight_number=flight["flight"]["iata"],
        airline_name=flight["airline"]["name"],
        aircraft_type=flight["aircraft"]["iata"],
        departure_time=flight["departure"]["estimated"],
        arrival_time=flight["arrival"]["estimated"],
        departure_city=flight["departure"]["airport"],
        arrival_city=flight["arrival"]["airport"]
    )

    return flight_data

def handle_oracle_request(request_data):
    flight_id = request_data.get("data")[0]

    try:
        flight_data = fetch_flight_data(flight_id)

        result = {
            "flight_status": flight_data.flight_status,
            "departure_city": flight_data.departure_city,
            "arrival_city": flight_data.arrival_city,
            "flight_number": flight_data.flight_number,
            "airline_name": flight_data.airline_name,
            "aircraft_type": flight_data.aircraft_type,
            "departure_time": flight_data.departure_time,
            "arrival_time": flight_data.arrival_time
        }

        response = {
            "result": result
        }

        return response

    except Exception as e:
        return {"error": str(e)}

request_data = {
    "data": ["AA100"]  # Example flight ID (American Airlines Flight 100)
}

response = handle_oracle_request(request_data)
print(response)
